src/Raven.Server/Documents/Indexes/MapReduce/AutoMapReduceIndex.cs

```diff
@@ -197,66 +187,81 @@ public override IQueryResultRetriever GetQueryResultRetriever(DocumentsOperation
 
-            var existingEntry = GetMapEntryForDocument(documentKey);
+            long id = -1;
 
-            long storageId;
-            if (existingEntry == null)
-            {
-                storageId = _mapReduceWorkContext.MapEntriesTable.Insert(tvb);
-            }
-            else
+            if (documentMapEntries.NumberOfEntries > 0)
             {
-                if (existingEntry.ReduceKeyHash != reduceKeyHash)
+                var mapEntries = GetMapEntriesForDocument(documentKey, documentMapEntries);
+
+                if (mapEntries.Count == 1 && mapEntries[0].ReduceKeyHash == reduceKeyHash)
+                {
+                    // update of existing entry, reduce key remained the same - we are going to overwrite the map result only
+                    id = mapEntries[0].Id;
+                }
+                else
                 {
-                    // reduce key changed - need to remove record in related reduce tree
+                    foreach (var mapEntry in mapEntries)
+                    {
+                        var previousState = GetReduceKeyState(mapEntry.ReduceKeyHash, indexContext, create: false);
+
+                        fixed (long* ptr = &mapEntry.Id)
+                            previousState.Tree.Delete(new Slice((byte*)ptr, sizeof(long)));
 
-                    var previousState = GetReduceKeyState(existingEntry.ReduceKeyHash, indexContext, create: false);
-                    var previousStorageId = existingEntry.StorageId;
-                    previousState.Tree.Delete(new Slice((byte*)&previousStorageId, sizeof(long)));
+                        documentMapEntries.Delete(mapEntry.Id);
+
+                        _mapReduceWorkContext.EntryDeleted(mapEntry);
+                    }
                 }
-                
-                storageId = _mapReduceWorkContext.MapEntriesTable.Update(existingEntry.StorageId, tvb);
             }
-           
-            state.Tree.DirectAdd(new Slice((byte*)&storageId, sizeof(long)), mappedResult.Size);
+
+            if (id == -1)
+            {
+                id = _mapReduceWorkContext.GetNextIdentifier();
+                documentMapEntries.Add(id, new Slice((byte*)&reduceKeyHash, sizeof(ulong)));
+            }
+
+            var pos = state.Tree.DirectAdd(new Slice((byte*)&id, sizeof(long)), mappedResult.Size);
+
+            mappedResult.CopyTo(pos);
         }
 
-        public unsafe MapEntry GetMapEntryForDocument(LazyStringValue documentKey)
+        public static unsafe List<MapEntry> GetMapEntriesForDocument(LazyStringValue documentKey, FixedSizeTree documentMapEntries)
         {
-            var documentKeySlice = new Slice(documentKey.Buffer, (ushort) documentKey.Size);
-
-            var seek = _mapReduceWorkContext.MapEntriesTable.SeekForwardFrom(_byDocumentKeysIndex, documentKeySlice).FirstOrDefault();
+            var entries = new List<MapEntry>((int)documentMapEntries.NumberOfEntries);
```

Comment:
We are processing this one at a time, it would likely be cheaper for us to allocate a single list here and use it for all documents.